src/ooni/ooni_test_impl.hpp

```diff
@@ -135,118 +138,61 @@ class OoniTestImpl : public mk::NetTest, public mk::NonCopyable,
         return filename;
     }
 
+    // Functions that derived classes SHOULD override
   protected:
     virtual void setup(std::string) {}
-    virtual void setup() {}
-
     virtual void teardown(std::string) {}
-    virtual void teardown() {}
-
-    virtual void main(Settings, std::function<void(report::Entry)> &&cb) {
-        reactor->call_later(1.25, [cb]() {
-            report::Entry entry;
-            cb(entry);
-        });
-    }
-
-    virtual void main(std::string, Settings,
-                      std::function<void(report::Entry)> &&cb) {
-        reactor->call_later(1.25, [cb]() {
-            report::Entry entry;
-            cb(entry);
-        });
-    }
-
-    void validate_input_filepath() {
-        if (input_filepath == "") {
-            throw InputFileRequired("An input file is required!");
-        }
-
-        struct stat buffer;
-        if (stat(input_filepath.c_str(), &buffer) != 0) {
-            throw InputFileDoesNotExist(input_filepath + " does not exist");
-        }
+    virtual void main(std::string, Settings, Callback<report::Entry> cb) {
+        reactor->call_soon([=]() { cb(report::Entry{}); });
     }
```

Comment:
Simplified: one unique main(). Input file path validation inline in `begin()`. Just one version of setup and teardown.